Profile API Support

huylenq opened the pull request:
Main code changes are in the `com.treasuredata.android.cdp` package.

There's a lot of noises due to 132f909, it will be cleaner after #40 is merged to master: https://github.com/treasure-data/td-android-sdk/compare/restructure...profile-api

tung-vu commented on test-host/README.md:
```diff
@@ -0,0 +1,3 @@
+To be used a host runtime for unit tests that involve platform API.
```
I haven't found the sdk README get updated with new Profile section :)

huylenq replied:
On my Todo

tung-vu commented on src/main/java/com/treasuredata/android/cdp/SegmentResultsOfJSON.java:
```diff
@@ -0,0 +1,86 @@
+    }
+
+    /**
+     * Error response, expect to be in the form of <code>{"error":..., "message":... }</code>
```
It is `<code>{"error":..., "message":..., "status":....}</code>`

tung-vu commented on src/main/java/com/treasuredata/android/cdp/SegmentResultsOfException.java:
```diff
@@ -0,0 +1,15 @@
+package com.treasuredata.android.cdp;
+
+public class SegmentResultsOfException extends SegmentsResult {
```
Should this SegmentResultsOfException and SegmentsResultOfJSON not be public?

tung-vu replied:
Also, can we generalize SegmentResultsOfException and  SegmentsResultsOfJSONObject with error format to something like CDPAPIException and CDPResponseException? I think these two case have nothing to do with SegmentsResults. The other two cases (SegmentsResultsOfJSONArray and SegmentsResultsOfJSONObject without an error format) can be extended from a new abstract class like CDPResponseJSON.
This would help tremendously future new api (for example fetchRealtimeUserSegments).
Of course we can always refactor later, it's up to you :) Just my 2 cents

huylenq replied:
I renamed to ErrorLookupResult and JSONObject/ArrayLookupResult, since the old terms sound confusing, there's not always segments data there.
The JSONLookupResult or ErrorLookupResult is not meant to be exceptions, they all (including JSONArrayLookupResult) are just intermediate encapsulation to able to know to invoke the callbacks. They hide the logic of how the callbacks get handled (either onSuccess or onCallback get called, although it's pretty obvious in the ErrorLookupResult name). The suggested way expose or at least imply that logic outside.

tung-vu commented on src/main/java/com/treasuredata/android/cdp/CDPAPIException.java:
```diff
@@ -0,0 +1,65 @@
+            if (json.has("error")) error = json.getString("error");
+            if (json.has("message")) message = json.getString("message");
+        } catch (JSONException e) {
+            // Since we checked,
```
Unfinished comment

tung-vu commented on src/main/java/com/treasuredata/android/cdp/CDPAPIException.java:
```diff
@@ -0,0 +1,65 @@
+        if (error.isEmpty() && message.isEmpty()) {
+            // Hopefully doesn't happen, but if the received JSON Object contains unexpected schema,
+            // then use the entire JSON as the exception message.
+            return new CDPAPIException(statusCode, null, json.toString());
```
Shouldn't the status of the json response here instead of http status?

huylenq commented on src/main/java/com/treasuredata/android/cdp/CDPClient.java:
```diff
@@ -0,0 +1,142 @@
+    }
+
+    /**
+     * @param profileAPITokens that are defined on TreasureData
```
Okay, `ProfilesTokens` then? Unlike iOS this could be changed anytime without breaking anything anyway.

huylenq commented on src/test/java/com/treasuredata/android/cdp/SegmentsResultTest.java:
```diff
@@ -0,0 +1,225 @@
+    @Test
+    public void should_fail_and_parse_error_and_message_on_200_with_json_error_schema() throws Exception {
+        SegmentsResult
+                .create(200, "{" +
```
Kind of an extreme case, but I prefer returning the actual HTTP status code. Will update the test anyway.

huylenq commented:
Summary the changes from 1f478e5...ca031f9:

- Added `CDPClientTest` which is not currently runnable on CI (I just discovered CircleCI's Linux box could not run Android emulator). On local, this test could be triggered with `./gradlew connectedAndroidTest`.
- Added `CDPAPIException` tests
- Expose the feature to TreasureData
- Some other renames, null checks, small fixes...

kietdo360 commented on README.md:
````diff
@@ -429,6 +429,26 @@ If you want to use server side upload timestamp not only client device time that
+Lookup for profiles via [Profile API](https://support.treasuredata.com/hc/en-us/sections/360000271688-Profiles-API)
+
+```
+TreasureData.sharedInstance().setCDPEndpoint("https://cdp.in.treasuredata.com");
````
Let change actual URL to placeholder i.e <your cdp instance region url>

tung-vu commented on src/main/java/com/treasuredata/android/cdp/FetchUserSegmentsCallback.java:
```diff
@@ -0,0 +1,19 @@
+
+import java.util.List;
+
+public interface FetchUserSegmentsCallback {
```
IMHO, this should be JSONLookupCallback. With FetchUserSegmentsCallback extends JSONLookupCallback. Overridden onSuccess and onError will be perfect place to convert JSON object to ProfileImpl POJO.

huylenq replied:
FetchUserSegmentsCallback should not be a sub-case of JSONLookupCallback, it should be the other way around. Plus this is a public interface, and user codes should not concern about these.
FetchUserSegmentsCallback just means this is a callback interface for FetchUserSegments operation, nothing more.

tung-vu commented on src/main/java/com/treasuredata/android/cdp/SegmentResultsOfJSON.java:
```diff
@@ -0,0 +1,87 @@
+        }
+
+        @Override
+        void invoke(FetchUserSegmentsCallback callback) {
```
Should invoke JSONLookupCallback (I'm just suggesting here. Name's up to you). JSONArrayLookupResult should not have anything to do with FetchUserSegmentsCallback

huylenq replied:
I quote the above:
> FetchUserSegmentsCallback just means this is a callback interface for FetchUserSegments operation, nothing more.
